flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java

```diff
@@ -421,7 +423,27 @@ public void reDistributeInputChannelStates(TaskStateAssignment stateAssignment)
                 stateAssignment.oldState.get(stateAssignment.inputOperatorID);
         final List<List<InputChannelStateHandle>> inputOperatorState =
                 splitBySubtasks(inputState, OperatorSubtaskState::getInputChannelState);
-        if (inputState.getParallelism() == executionJobVertex.getParallelism()) {
+
+        boolean noNeedRescale =
+                stateAssignment.executionJobVertex.getJobVertex().getInputs().stream()
+                                .map(JobEdge::getDownstreamSubtaskStateMapper)
+                                .anyMatch(m -> !m.equals(SubtaskStateMapper.FULL))
+                        && stateAssignment.executionJobVertex.getInputs().stream()
+                                .map(IntermediateResult::getProducer)
+                                .map(vertexAssignments::get)
+                                .anyMatch(
+                                        taskStateAssignment -> {
+                                            final int oldParallelism =
+                                                    stateAssignment
+                                                            .oldState
+                                                            .get(stateAssignment.inputOperatorID)
+                                                            .getParallelism();
+                                            return oldParallelism
+                                                    == taskStateAssignment.executionJobVertex
+                                                            .getParallelism();
+                                        });
```

Comment:
Could you explain the logic behind this condition? Maybe both of it's parts separately?
```
                stateAssignment.executionJobVertex.getJobVertex().getInputs().stream()
                                .map(JobEdge::getDownstreamSubtaskStateMapper)
                                .anyMatch(m -> !m.equals(SubtaskStateMapper.FULL))
```
and 
                               
```
 stateAssignment.executionJobVertex.getInputs().stream()
                                .map(IntermediateResult::getProducer)
                                .map(vertexAssignments::get)
                                .anyMatch(
                                        taskStateAssignment -> {
                                            final int oldParallelism =
                                                    stateAssignment
                                                            .oldState
                                                            .get(stateAssignment.inputOperatorID)
                                                            .getParallelism();
                                            return oldParallelism
                                                    == taskStateAssignment.executionJobVertex
                                                            .getParallelism();
                                        }
```
?

Reply:
Separation is a good idea, ok.
The first part is to check if SubtaskStateMapper.FULL is used in any input (this mapper type always returns all I/O), the problem is reproduced only for this mapper type
The second part is to check if the parallelism of any previous operator has changed. if it has changed, it means that the number of outputs has changed.

Reply:
Thanks for the explanation.

> But I made little mistake in condition. I will fix it

Does this bug have a test coverage? I mean, either was there some test failing or have you added a new test to cover for a future regression?

Reply:
Maybe in `StateAssignmentOperationTest` create a unit test that has one `FULL` and one something else, and assert that the assigned states are as they should be?

Reply:
> Does this bug have a test coverage? I mean, either was there some test failing or have you added a new test to cover for a future regression?


Hey, sorry to bother you again, but the unit test that you have added still doesn't have test coverage. When I try running your previous version of the code:
```
        boolean noNeedRescale =
                stateAssignment.executionJobVertex.getJobVertex().getInputs().stream()
                                .map(JobEdge::getDownstreamSubtaskStateMapper)
                                .anyMatch(m -> !m.equals(SubtaskStateMapper.FULL))
                        && stateAssignment.executionJobVertex.getInputs().stream()
                                .map(IntermediateResult::getProducer)
                                .map(vertexAssignments::get)
                                .anyMatch(
                                        taskStateAssignment -> {
                                            final int oldParallelism =
                                                    stateAssignment
                                                            .oldState
                                                            .get(stateAssignment.inputOperatorID)
                                                            .getParallelism();
                                            return oldParallelism
                                                    == taskStateAssignment.executionJobVertex
                                                            .getParallelism();
                                        });

        if (inputState.getParallelism() == executionJobVertex.getParallelism() && !noNeedRescale) {
            stateAssignment.inputChannelStates.putAll(
                    toInstanceMap(stateAssignment.inputOperatorID, inputOperatorState));
            return;
        }
```

the tests in `StateAssignmentOperationTest` are still green.

Reply:
Thanks for the analysis.  I think you can add a test with incomplete FULL, you are right.  I hope I didn't make a mistake with the error note last time.

Reply:
Earlier, I misjudged the error of not noticing the negation in the condition. 
But by substituting `anyMatch` for `allMatch`, you can see that the test will stop passing.
Also I changed gateIdx from random generation to always zero (in `StateHandleDummyUtil#createNewInputChannelStateHandle`). This change have no affects for other tests. But I need zero for this test, because you need stability for the number of input gates